File: src/openprecedent/schemas/decision.py
from enum import StrEnum

from pydantic import BaseModel, ConfigDict, Field


class DecisionType(StrEnum):
    TASK_FRAME_DEFINED = "task_frame_defined"
    CONSTRAINT_ADOPTED = "constraint_adopted"
    SUCCESS_CRITERIA_SET = "success_criteria_set"
    CLARIFICATION_RESOLVED = "clarification_resolved"
    OPTION_REJECTED = "option_rejected"
    AUTHORITY_CONFIRMED = "authority_confirmed"


class DecisionExplanation(BaseModel):
    model_config = ConfigDict(extra="forbid")

    goal: str
    evidence: list[str] = Field(default_factory=list)
    constraints: list[str] = Field(default_factory=list)
    selection_reason: str
    result: str | None = None


class Decision(BaseModel):
    model_config = ConfigDict(extra="forbid")

    decision_id: str
    case_id: str
    decision_type: DecisionType
    title: str
    question: str
    chosen_action: str
    alternatives: list[str] = Field(default_factory=list)
    evidence_event_ids: list[str] = Field(default_factory=list)
    constraint_summary: str | None = None
    requires_human_confirmation: bool = False
    outcome: str | None = None
    confidence: float = 0.5
    explanation: DecisionExplanation
    sequence_no: int

File: rust/openprecedent-contracts/src/decision.rs
use serde::{Deserialize, Serialize};
use strum::{Display, EnumString};

#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize, Deserialize, Display, EnumString)]
#[serde(rename_all = "snake_case")]
#[strum(serialize_all = "snake_case")]
pub enum DecisionType {
    TaskFrameDefined,
    ConstraintAdopted,
    SuccessCriteriaSet,
    ClarificationResolved,
    OptionRejected,
    AuthorityConfirmed,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct DecisionExplanation {
    pub goal: String,
    pub evidence: Vec<String>,
    pub constraints: Vec<String>,
    pub selection_reason: String,
    pub result: Option<String>,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct Decision {
    pub decision_id: String,
    pub case_id: String,
    pub decision_type: DecisionType,
    pub title: String,
    pub question: String,
    pub chosen_action: String,
    pub alternatives: Vec<String>,
    pub evidence_event_ids: Vec<String>,
    pub constraint_summary: Option<String>,
    pub requires_human_confirmation: bool,
    pub outcome: Option<String>,
    pub confidence: f64,
    pub explanation: DecisionExplanation,
    pub sequence_no: i64,
}
